antarest/launcher/web.py

```diff
@@ -203,14 +205,44 @@ def get_load(
     @bp.get(
         "/launcher/_versions",
         tags=[APITag.launcher],
-        summary="Get list of supported study version for all configured launchers",
-        response_model=Dict[str, List[str]],
+        summary="Get list of supported solver versions",
+        response_model=List[str],
     )
-    def get_versions(
-        current_user: JWTUser = Depends(auth.get_current_user),
-    ) -> Any:
-        params = RequestParameters(user=current_user)
-        logger.info(f"Fetching version list")
-        return service.get_versions(params=params)
+    def get_solver_versions(
+        solver: Optional[str] = Query(
+            None,
+            examples={
+                "default solver": {
+                    "description": "Get the solver versions of the default configuration",
+                    "value": "",
+                },
+                "SLURM solver": {
+                    "description": "Get the solver versions of the SLURM server if available",
+                    "value": "slurm",
+                },
+                "Local solver": {
+                    "description": "Get the solver versions of the Local server if available",
+                    "value": "local",
+                },
+            },
+        )
+    ) -> List[str]:
+        """
+        Get list of supported solver versions defined in the configuration.
+
+        Args:
+        - `solver`: name of the configuration to read: "default", "slurm" or "local".
+        """
+        solver = solver or "local"
+        logger.info(
+            f"Fetching the list of solver versions for the '{solver}' configuration"
+        )
+        try:
+            return service.get_solver_versions(solver)
+        except ValueError as e:
+            raise HTTPException(
+                status_code=http.HTTPStatus.UNPROCESSABLE_ENTITY,
+                detail=str(e),
+            )
```

Comment:
redundant error check with the service function `get_solver_versions`,we can move the check for an invalid solver configuration from the service to the route handler. The route handler is the appropriate place to validate user input, declaring an exception with appropriate message:

we can also be more specific and check against the params allowed only:

```python
 if solver not in ["default", "slurm", "local"]:
            raise InvalidSolverConfig
```

Finnaly we would have this:

```python
      solver = solver or "default"

        if solver not in ["default", "slurm", "local"]:
            raise InvalidSolverConfig
        logger.info(
            f"Fetching the list of solver versions for the '{solver}' configuration"
        )

        return service.get_solver_versions(solver)
```

Reply:
I think an exception called `UnknownSolverConfig` is more appropriate, because the error message should be something like that: "Unknown solver configuration: '{solver}'". Do not forget to pass the `solver` value as a parameter to the exception.

I suggest this:

```python
class UnknownSolverConfig(HTTPException):
    """
    Exception raised during solver versions retrieval when
    the name of the launcher is not "default", "slurm" or "local".
    """

    def __init__(self, solver: str) -> None:
        super().__init__(
            http.HTTPStatus.UNPROCESSABLE_ENTITY,
            f"Unknown solver configuration: '{solver}'",
        )

solver = solver or "local"
logger.info(
    f"Fetching the list of solver versions for the '{solver}' configuration"
)
if solver not in {"default", "slurm", "local"}:
    raise UnknownSolverConfig(solver)
return service.get_solver_versions(solver)
```

Reply:
```python
def get_solver_versions(self, solver: str) -> List[str]:
    """
    Fetch the list of solver versions from the configuration.

    Args:
        solver: name of the configuration to read: "default", "slurm" or "local".

    Returns:
        The list of solver versions.
        This list is empty if the configuration is not available.

    Raises:
        KeyError: if the configuration is not "default", "slurm" or "local".
    """
    local_config = self.config.launcher.local
    slurm_config = self.config.launcher.slurm
    default_config = self.config.launcher.default
    versions_map = {
        "local": sorted(local_config.binaries) if local_config else [],
        "slurm": sorted(slurm_config.antares_versions_on_remote_server)
        if slurm_config
        else [],
    }
    versions_map["default"] = versions_map[default_config]
    return versions_map[solver]
```